cmd/bbolt/command_surgery_meta.go

```diff
@@ -0,0 +1,317 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	"go.etcd.io/bbolt/internal/common"
+)
+
+func newSurgeryMetaCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "meta <subcommand>",
+		Short: "meta page related surgery commands",
+	}
+
+	cmd.AddCommand(newSurgeryMetaValidateCommand())
+	cmd.AddCommand(newSurgeryMetaUpdateCommand())
+
+	return cmd
+}
+
+type surgeryMetaValidateOptions struct {
+	pageSize uint32
+}
+
+func (o *surgeryMetaValidateOptions) AddFlags(fs *pflag.FlagSet) {
+	fs.Uint32VarP(&o.pageSize, "page-size", "", o.pageSize, "page size (used to locate the second meta page, use the os default page size if zero")
+}
+
+func newSurgeryMetaValidateCommand() *cobra.Command {
+	var o surgeryMetaValidateOptions
+	metaValidateCmd := &cobra.Command{
+		Use:   "validate <bbolt-file> [options]",
+		Short: "Validate both meta pages",
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) == 0 {
+				return errors.New("db file path not provided")
+			}
+			if len(args) > 1 {
+				return errors.New("too many arguments")
+			}
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return surgeryMetaValidateFunc(args[0], o)
+		},
+	}
+	o.AddFlags(metaValidateCmd.Flags())
+	return metaValidateCmd
+}
+
+func surgeryMetaValidateFunc(srcDBPath string, cfg surgeryMetaValidateOptions) error {
+	if _, err := checkSourceDBPath(srcDBPath); err != nil {
+		return err
+	}
+
+	pageSize := cfg.pageSize
+	if pageSize == 0 {
+		pageSize = uint32(common.DefaultPageSize)
+	}
+
+	var (
+		m0ValidateResult error
+		m1ValidateResult error
+	)
+
+	// validate meta page 0
+	m0, _, err := ReadMetaPageAt(srcDBPath, 0, pageSize)
+	if err != nil {
+		return fmt.Errorf("read meta page 0 failed: %w", err)
+	}
+	m0ValidateResult = m0.Validate()
+
+	if m0ValidateResult == nil && pageSize != m0.PageSize() {
+		fmt.Fprintf(os.Stdout, "WARNING: The provided page size (%d) isn't consistent with the page size (%d) in the first valid meta page\n", pageSize, m0.PageSize())
```

Comment:
If the first meta page can be read and valid, we should ignore the pageSize set by user and use the first meta's page size.

Reply:
Sometimes, the meta page's checksum is correct, but some fields's value may be wrong, such as https://github.com/etcd-io/bbolt/issues/537.